### PR TITLE
Integrate TensorRT EP libs into existing GPU Nuget Package (outdated)

### PR DESCRIPTION
**Pull request by chilo-ms**
Description: Integrate TensorRT EP shared libraries into existing GPU Nuget Package as part of the goal to create combined gpu package.

We plan to do it with two approaches:

1. Create another Pipeline’s job to separately build TRT EP. And then extract TRT EP shared library and put it into GPU nuget package.
3. Add additional option --use_tensorrt in current build command to include building TRT EP library.

This PR is approach # 1. 
We plan to adopt approach # 1 first and then later we can adopt approach # 2.

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml [hunk 1]
```diff
@@ -133,7 +156,7 @@ jobs:
...
-          /bin/bash -c "/opt/python/cp37-cp37m/bin/python3 /onnxruntime_src/tools/ci_build/build.py --build_dir /build --config Release --skip_submodule_sync  --parallel --build_shared_lib --use_cuda --cuda_version=11.1 --cuda_home=/usr/local/cuda-11.1 --cudnn_home=/usr/local/cuda-11.1 --enable_onnx_tests --cmake_extra_defines CMAKE_CUDA_HOST_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/cc 'CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80' && cd /build/Release && make install DESTDIR=/build/linux-x64"
+          /bin/bash -c "/opt/python/cp37-cp37m/bin/python3 /onnxruntime_src/tools/ci_build/build.py --build_dir /build --config Release --skip_submodule_sync  --parallel --build_shared_lib --use_cuda --cuda_version=11.1 --cuda_home=/usr/local/cuda-11.1 --cudnn_home=/usr/local/cuda-11.1 --enable_onnx_tests --skip_tests --cmake_extra_defines CMAKE_CUDA_HOST_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/cc 'CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80' && cd /build/Release && make install DESTDIR=/build/linux-x64"
```
Comment (jywu-msft) [inline tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml:159]: looks like the change was to add skip_tests? why?
and should we align on same cuda verison 11.4?

File: tools/nuget/generate_nuspec_for_native_nuget.py [hunk 2]
```diff
@@ -181,6 +181,7 @@ def generate_files(list, args):
     is_cpu_package = args.package_name in ['Microsoft.ML.OnnxRuntime', 'Microsoft.ML.OnnxRuntime.OpenMP']
     is_mklml_package = args.package_name == 'Microsoft.ML.OnnxRuntime.MKLML'
     is_cuda_gpu_package = args.package_name == 'Microsoft.ML.OnnxRuntime.Gpu'
+    is_tensorrt_package = args.package_name == 'Microsoft.ML.OnnxRuntime.TensorRT'
```
Comment (jywu-msft) [inline tools/nuget/generate_nuspec_for_native_nuget.py:184]: I was thinking there won't be such a thing as a TensorRT package.
i.e. The ORT Gpu package will now also include TensorRT EP shared lib.

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml (hunk 1 above)
Comment (chilo-ms) [reply tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml:159]: Sorry, I didn't remove skip_tests flag. It's for testing only, will remove this flag.

File: tools/nuget/generate_nuspec_for_native_nuget.py (hunk 2 above)
Comment (chilo-ms) [reply tools/nuget/generate_nuspec_for_native_nuget.py:184]: It's just a intermediate package name not the final package name.

'Microsoft.ML.OnnxRuntime.TensorRT' is being used as intermediate name and this name is also the name of intermediate package:
https://github.com/microsoft/onnxruntime/blob/3303a832063ecbe8febfce366ffd105374879a16/tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml#L21
https://github.com/microsoft/onnxruntime/blob/tensorrt_ep_nuget_package_2/csharp/OnnxRuntime.CSharp.proj#L129

Comment (jywu-msft) [reply tools/nuget/generate_nuspec_for_native_nuget.py:184]: ok thanks

**Review by jywu-msft — DISMISSED** [review]

File: csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests/runtest.bat [hunk 3]
```diff
@@ -37,16 +38,21 @@ IF EXIST test\Microsoft.ML.OnnxRuntime.EndToEndTests\packages RMDIR /S /Q test\M
...
 IF NOT errorlevel 0 (
     @echo "Failed to restore nuget packages for the test project"
     EXIT 1
 )
 
-%dn% test test\Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj --no-restore
+%dn% list test\Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj package
+dir test\Microsoft.ML.OnnxRuntime.EndToEndTests\packages\
+
+%dn% test test\Microsoft.ML.OnnxRuntime.EndToEndTests\Microsoft.ML.OnnxRuntime.EndToEndTests.csproj --no-restore /p:DefineConstants="USE_TENSORRT"
```
Comment (jywu-msft) [inline csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests/runtest.bat:55]: @snnn pointed out that /p:DefineConstants="USE_TENSORRT" impacts all cases whether TRT is there or not.

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml [hunk 4]
```diff
@@ -5,19 +5,42 @@ parameters:
 jobs:
 - template: ../../templates/win-ci-2019.yml
   parameters:
-    AgentPool : 'onnxruntime-gpu-winbuild'
+    AgentPool : 'onnxruntime-gpu-tensorrt8-winbuild'
+    ArtifactName: 'drop-nuget-tensorrt'
+    JobName: 'Windows_CI_GPU_TENSORRT_Dev'
+    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --enable_onnx_tests --use_telemetry --cmake_generator "Visual Studio 16 2019" --use_tensorrt --tensorrt_home="C:\local\TensorRT-8.0.1.6.Windows10.x86_64.cuda-11.3.cudnn8.2" --cuda_version=11.4 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4" --cudnn_home="C:\local\cudnn-11.4-windows-x64-v8.2.2.26\cuda" --cmake_extra_defines "CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80"
```
Comment (jywu-msft) [inline tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml:11]: i think we need to remove 50 from CUDA_ARCHITECTURE 
52 covers m60   , 50 is no longer supported i believe.

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml [hunk 5]
```diff
@@ -5,19 +5,42 @@ parameters:
...
+     copy $(Build.BinariesDirectory)\RelWithDebInfo\RelWithDebInfo\custom_op_library.* $(Build.ArtifactStagingDirectory)\testdata
+
+- template: ../../templates/win-ci-2019.yml
+  parameters:
+    AgentPool : 'onnxruntime-gpu-tensorrt8-winbuild'
     ArtifactName: 'drop-nuget'
     JobName: 'Windows_CI_GPU_CUDA_Dev'
-    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --enable_onnx_tests --use_telemetry --cmake_generator "Visual Studio 16 2019" --use_cuda --cuda_version=11.1 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1" --cudnn_home="C:\local\cudnn-11.1-windows-x64-v8.0.5.39\cuda" --cmake_extra_defines "CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80"
+    BuildCommand: --build_dir $(Build.BinariesDirectory) --skip_submodule_sync --build_shared_lib --enable_onnx_tests --use_telemetry --cmake_generator "Visual Studio 16 2019" --use_cuda --cuda_version=11.4 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4" --cudnn_home="C:\local\cudnn-11.4-windows-x64-v8.2.2.26\cuda" --cmake_extra_defines "CMAKE_CUDA_ARCHITECTURES=37;50;52;60;61;70;75;80"
```
Comment (jywu-msft) [inline tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml:34]: remove 50 here too

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml (hunk 1 above)
Comment (jywu-msft) [reply tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml]: what about the cuda 11.1 part? does that have other dependencies (and needs to be done in a separate PR) ?

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml (hunk 4 above)
Comment (chilo-ms) [reply tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml]: Removed.

File: tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml (hunk 5 above)
Comment (chilo-ms) [reply tools/ci_build/github/azure-pipelines/nuget/templates/gpu.yml]: Removed.

File: csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests/runtest.bat (hunk 3 above)
Comment (chilo-ms) [reply csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests/runtest.bat:56]: Modified the code to add /p:DefineConstants="USE_TENSORRT" only if it's GPU package.

**Comment by chilo-ms** [comment]
Please see the other PR #8727 which has the latest implementation.

**Pull request closed without merging**